### DISCUSSION TITLE
How should we name web map tiles generator?

### DISCUSSION THREAD

**Discussion by jtmiclat** [thinkingmachines/geowrangler#90] [General]
We created a Grid Generator named BingTileGridGenerator in #82. This was based on https://docs.microsoft.com/en-us/bingmaps/articles/bing-maps-tile-system. After further research,  the name might not make 100% sense. This is because bing tiles use the same tiling as most web mapping software (gmaps, mapbox). This tiling is an instance of [Tile Matrix](
 http://docs.opengeospatial.org/is/17-083r2/17-083r2.html) . Specifically a tile matrix for the EPGS:3857 projection. What makes Bing Maps unique is that they use Quadkeys as a unique identifier instead of the traditional x,y,z used by mapbox or gmaps. 
 
 Some possible namings and API 
 
 `TileMatrixGridGenerator` - where you can specify projection and target id format (either x,y,z or quadkey if possible)
 `WebMapGridGenerator` - where you can specify target id format (either x,y,z or quadkey if possible)
 `QuadkeyGridGenerator` - Still a misnomer since quadkeys are just ids and note actually say anything about the shape

**Comment by joshuacortez** [comment]
If we were to go with `TileMatrixGridGenerator`, how easy would that be to implement for arbitrary projections? Also I wonder if there a way to identify cell adjacency via the quadkey. It can be useful for downstream applications like cell clustering. From the [Bing Maps article](https://docs.microsoft.com/en-us/bingmaps/articles/bing-maps-tile-system), it says 

> Finally, quadkeys provide a one-dimensional index key that usually preserves the proximity of tiles in XY space. In other words, two tiles that have nearby XY coordinates usually have quadkeys that are relatively close together

 Is it really just _usually_, not always? 😅  Having x,y,z (even if it's just `WebMapGridGenerator`) could be useful if it can identify adjacency. 😄

**Reply by jtmiclat** [reply]
> how easy would that be to implement for arbitrary projections? 

The underlying library we use is morecantile. It has built-in projections
```
    CanadianNAD83_LCC: Lambert conformal conic NAD83 for Canada - EPSG:3978
    EuropeanETRS89_LAEAQuad: ETRS89-extended / LAEA Europe - EPGS:3035
    LINZAntarticaMapTilegrid: LINZ Antarctic Map Tile Grid (Ross Sea Region) - EPSG:5482
    NZTM2000: LINZ NZTM2000 Map Tile Grid - EPSG:2193
    NZTM2000Quad: LINZ NZTM2000 Map Tile Grid - EPSG:2193
    UPSAntarcticWGS84Quad: Universal Polar Stereographic WGS 84 Quad for Antarctic - EPSG:5042
    UPSArcticWGS84Quad: Universal Polar Stereographic WGS 84 Quad for Arctic - EPSG:5041
    UTM31WGS84Quad: Example of UTM grid - EPSG:32631
    WebMercatorQuad: Spherical Mercator - EPGS:3857 (default grid for Web Mercator based maps)
    WGS1984Quad: EPSG:4326 for the World - EPGS:4326 (WGS84)
    WorldCRS84Quad: CRS84 for the World
    WorldMercatorWGS84Quad: Elliptical Mercator projection - EPGS:3395
```
and the ability to add arbitrary projections. https://developmentseed.org/morecantile/usage/#define-custom-grid 


>  Also I wonder if there a way to identify cell adjacency via the quad key

it might just be easier to covert quadkey  to x, y, z and determine the adjacentcy

> Is it really just usually, not always?

I feel like the edge cases here are the boarders of Web Mercator (180 lon, -90/90 lat) since from max it becomes zero

**Reply by joshuacortez** [reply]
Oh nice, converting quadkey to x, y, z seems good

**Reply by joshuacortez** [reply]
We can convert quadkey to x,y using [a function like here](https://gis.stackexchange.com/questions/206202/how-can-i-convert-bings-quadtree-tile-addresses-to-zxy-tile-addresses-in-pyth) so we can easily do tile clustering. We can add a method or argument to `BingTileGridGenerator` that adds x,y coordinate columns, then we can use the resulting geodataframe with the existing `TileClustering` class.

**Comment by joshuacortez** [comment]
Closing since since [both Square Grid Generator and Bing Tile Grid Generator](https://geowrangler.thinkingmachin.es/grids.html) (as well as their fast versions) can return x,y columns.
